### PR TITLE
Wjordan213  handle ecs injected env vars

### PR DESCRIPTION
**Pull request by wjordan213**
still need to test things out with a local credentials container but I think it's worth starting the review process now

### CODE REVIEW

File: containers/demo/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,14 @@
...
+
+# 2. Copy the files in your machine to the Docker image
+COPY ./ ./
+
+# Build your program for release
+ADD /containers/demo/vault-ca.crt /opt/vault/certs/
+
+RUN cargo build --verbose
+
+RUN ls target
```
Comment (zellio) [inline containers/demo/Dockerfile:11]: This action doesn't do anything, please remove.

File: containers/demo/Dockerfile [hunk 2]
```diff
@@ -0,0 +1,14 @@
...
+
+# Run the binary
+CMD ["./target/debug/vault-env-exec", "spark:spark", "--", "echo", "hi"]
```
Comment (zellio) [inline containers/demo/Dockerfile:14]: Divide into:

```dockerfile
ENTRYPOINT ["./target/debug/vault-env-exec"]
CMD [...]
```

File: containers/demo/vault-ca.crt [hunk 3]
```diff
@@ -0,0 +1,20 @@
+-----BEGIN CERTIFICATE-----
```
Comment (zellio) [inline containers/demo/vault-ca.crt:1]: I don't know that we want to put our internal cert auth into a public repo in this way. It's not a secret value strictly speaking, but it is messy.

File: docker-compose.override.yml [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+      networks:
+          credentials_network:
+              ipv4_address: "169.254.170.3"
+      environment:
+        AWS_DEFAULT_REGION: "us-east-1"
+        AWS_CONTAINER_CREDENTIALS_RELATIVE_URI: "/creds"
+        VAULT_ADDR: "https://vault.trialspark.com:8200"
+        VAULT_SECURITY_HEADER: "vault.trialspark.com"
+        VAULT_ROLE: "spark-infra-test-reader"
+        VAULT_CACERT: "/opt/vault/certs/vault-ca.crt"
```
Comment (zellio) [inline docker-compose.override.yml:53]: This is super internal specific and we probably want to remove in favour of a more local testing schema

File: src/main.rs [hunk 5]
```diff
@@ -379,6 +385,47 @@ fn exec(command: String, args: &Vec<String>, env: &Vec<CString>) {
...
+    secret_access_key: String,
+    token: String,
+}
+
+#[tokio::main]
+async fn fetch_aws_credentials(
+    relative_uri: &String,
+    logger: &slog::Logger,
+) -> Result<(String, String, Option<String>), Box<dyn error::Error>> {
+    let aws_container_credentials_uri = format!("http://169.254.170.2{}", relative_uri);
```
Comment (zellio) [inline src/main.rs:403]: Break this ip-address out into a value passed in as an argument with default value for testing purposes.

**Review by zellio — CHANGES_REQUESTED** [review]

File: containers/demo/vault-ca.crt (hunk 3 above)
Comment (wjordan213) [reply containers/demo/vault-ca.crt:1]: what would the alternative be

Comment (wjordan213) [reply containers/demo/vault-ca.crt]: i .gitignored this

File: containers/demo/Dockerfile [hunk 6]
```diff
@@ -0,0 +1,16 @@
+FROM rust:1.61
+
+# 2. Copy the files in your machine to the Docker image
+COPY ./ ./
+
+# Build your program for release
+ADD /containers/demo/vault-ca.crt /certs/
+
+RUN set -ex; \
```
Comment (zellio) [inline containers/demo/Dockerfile:9]: If there is only on command in this sequence you don't need `set -ex`

File: containers/demo/Dockerfile [hunk 7]
```diff
@@ -0,0 +1,16 @@
+FROM rust:1.61
+
+# 2. Copy the files in your machine to the Docker image
+COPY ./ ./
+
+# Build your program for release
+ADD /containers/demo/vault-ca.crt /certs/
```
Comment (zellio) [inline containers/demo/Dockerfile:7]: rename folder from demo to certs and add a .gitignore file to it with * so the directory is always present

File: .gitignore [hunk 8]
```diff
@@ -1 +1,2 @@
 /target
+containers/demo/vault-ca.crt
```
Comment (zellio) [inline .gitignore:2]: Move this ignore to the directory specific case

File: docker-compose.yml [hunk 9]
```diff
@@ -0,0 +1,10 @@
+version: "2"
```
Comment (zellio) [inline docker-compose.yml:1]: Why split these across two different files?

**Review by zellio — CHANGES_REQUESTED** [review]

File: docker-compose.yml (hunk 9 above)
Comment (wjordan213) [reply docker-compose.yml:1]: that was something the AWS guide specifically recommended. no idea why

Comment (zellio) [reply docker-compose.yml]: lol, interesting. I am generally distrustful of AWS docs

File: docker-compose.yml [hunk 10]
```diff
@@ -0,0 +1,10 @@
+version: "3"
```
Comment (zellio) [inline docker-compose.yml:1]: This file adds no value, please delete and rename the override file to docker-compose.yml

**Review by zellio — APPROVED** [review]

**Pull request merged**